### ISSUE TITLE
Waves exchange initialization issue

### ISSUE THREAD

**Issue by rmcxm** [ccxt/ccxt#7118]
I am unable to use waves apis. Is instantiation correct?

- OS: fedora 30
- Programming Language version: python 3.7
- CCXT version: 1.29.67 (latest)

```
exchange = ccxt.wavesexchange({'options': { }, 'apiKey':'waves private key used here', 'secret':'waves access token used here'   })
exchange.fetch_order_book("BTCUSDT",limit=1)
```

```
Traceback (most recent call last):
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/urllib3/response.py", line 397, in _error_catcher
    yield
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/urllib3/response.py", line 707, in read_chunked
    chunk = self._handle_chunk(amt)
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/urllib3/response.py", line 662, in _handle_chunk
    returned_chunk = self._fp._safe_read(self.chunk_left)
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/http/client.py", line 620, in _safe_read
    chunk = self.fp.read(min(amt, MAXAMOUNT))
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/socket.py", line 589, in readinto
    return self._sock.recv_into(b)
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/ssl.py", line 1071, in recv_into
    return self.read(nbytes, buffer)
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/ssl.py", line 929, in read
    return self._sslobj.read(len, buffer)
socket.timeout: The read operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/requests/models.py", line 750, in generate
    for chunk in self.raw.stream(chunk_size, decode_content=True):
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/urllib3/response.py", line 527, in stream
    for line in self.read_chunked(amt, decode_content=decode_content):
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/urllib3/response.py", line 732, in read_chunked
    self._original_response.close()
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/urllib3/response.py", line 402, in _error_catcher
    raise ReadTimeoutError(self._pool, None, 'Read timed out.')
urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='matcher.waves.exchange', port=443): Read timed out.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/ccxt/base/exchange.py", line 561, in fetch
    verify=self.verify
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/requests/sessions.py", line 686, in send
    r.content
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/requests/models.py", line 828, in content
    self._content = b''.join(self.iter_content(CONTENT_CHUNK_SIZE)) or b''
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/requests/models.py", line 757, in generate
    raise ConnectionError(e)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='matcher.waves.exchange', port=443): Read timed out.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/ccxt/wavesexchange.py", line 388, in fetch_order_book
    self.load_markets()
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/ccxt/base/exchange.py", line 1321, in load_markets
    markets = self.fetch_markets(params)
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/ccxt/wavesexchange.py", line 349, in fetch_markets
    response = self.matcherGetMatcherOrderbook(params)
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/ccxt/base/exchange.py", line 462, in inner
    return entry(_self, **inner_kwargs)
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/ccxt/base/exchange.py", line 485, in request
    return self.fetch2(path, api, method, params, headers, body)
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/ccxt/base/exchange.py", line 481, in fetch2
    return self.fetch(request['url'], request['method'], request['headers'], request['body'])
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/ccxt/base/exchange.py", line 601, in fetch
    raise ExchangeError(method + ' ' + url + ' ' + error_string)
ccxt.base.errors.ExchangeError: GET https://matcher.waves.exchange/matcher/orderbook HTTPSConnectionPool(host='matcher.waves.exchange', port=443): Read timed out.

```

**Comment by kroitor** [comment]
Hi!

Can you please try this code and let us know if it throws the same exception:

```Python
exchange = ccxt.wavesexchange({
    'urls': {
        'api': {
            'matcher': 'http://matcher.waves.exchange',  # add this override
        },
    },
})
print(exchange.fetch_order_book("BTC/USDT",limit=1))  # mind the symbol here 
```

↑ If that works, we will add a quickfix for it.

**Comment by rmcxm** [comment]
It works!

But it asks for argument limit as string not int. When I gave limit = "1" in args, it dumped too many asks and bids from orderbook instead of just 1.

`ccxt.base.errors.ExchangeError: GET http://matcher.waves.exchange/matcher/orderbook/WAVES/8LQW8f7P5d5PZM7GtZEBgaqRPGSzS3DfPuiXrURJ4AJS Value for header {depth: 1} must be of type str or bytes, not <class 'int'>
`

`>>> exchange.fetch_order_book("WAVES/BTC",limit="1")
{'bids': [[0.0001272, 10107.68977986], [0.00012702, 502.44641745], [0.00012633, 74.0], [0.00012624, 133.57536082], [0.00012616, 81.0], [0.00012602, 88.0], [0.00012585, 221.0], [0.00012574, 267.15072164], [0.0001257, 382.20114649], [0.00012561, 20.0], [0.00012553, 475.90799494], [0.00012542, 8.5305466], [0.0001252, 1574.9869145], [0.0001251, 0.11], [0.00012489, 157.27823158], [0.00012481, 568.79725351], [0.0001246, 1137.39382022], [0.00012458, 859.92103996], [0.00012449, 20.0], [0.00012426, 270.94268527], [0.0001241, 372.65423046], [0.00012401, 5.0], [0.00012394, 925.60967796], [0.00012392, 0.33], [0.00012362, 699.50027019], [0.00012359, 6.40081601], [0.00012338, 20.0], [0.00012331, 805.93927625], [0.0001231, 150.0], [0.00012301, 203.23550931], [0.000123, 8.45639318], [0.00012298, 428.13974676], [0.00012296, 6.41124277], [0.00012266, 232.76950046], [0.00012228, 20.0], [0.00012222, 5.0], [0.0001222, 4380.22], [0.0001221, 9.3125243], [0.00012209, 10.02006978], [0.000122, 7.79794878], [0.00012179, 2037.21627391], [0.00012177, 853.19588026], [0.00012152, 8.89663212], [0.00012119, 20.0], [0.00012111, 94.0], [0.0001211, 0.125], [0.00012104, 20.53105619], [0.000121, 8.09829176], [0.00012095, 8.2394964], [0.00012092, 0.55], [0.0001208, 205.2545156], [0.00012065, 0.61840034], [0.00012061, 0.22], [0.0001205, 6.91593361], [0.00012038, 8.59471254], [0.00012018, 0.77], [0.00012011, 20.0], [0.0001201, 1.07410491], [0.00012009, 0.12], [0.00012002, 465.61723046], [0.00012, 2878.74724221], [0.00011982, 8.59454566], [0.00011926, 8.2071342], [0.00011919, 2.46], [0.00011918, 0.99], [0.00011909, 10.0], [0.00011904, 20.0], [0.00011888, 28.68438761], [0.0001187, 8.207217], [0.00011814, 7.86793073], [0.00011809, 75.78911], [0.00011801, 1.99644098], [0.000118, 100.87744496], [0.00011761, 0.33], [0.00011759, 8.20732613], [0.0001175, 100.0], [0.00011733, 105.08829796], [0.00011715, 5.0], [0.00011711, 0.33], [0.0001171, 2137.59453458], [0.00011709, 0.255], [0.00011707, 160.16058768], [0.00011704, 7.28616915], [0.000117, 13.09129906], [0.00011697, 115.33889031], [0.00011649, 8.97966375], [0.00011646, 9.0], [0.00011635, 20.0], [0.00011632, 283.0], [0.0001162, 2.0], [0.000116, 1012.05806748], [0.00011595, 8.27085969], [0.00011592, 1.0], [0.00011575, 285.38548596], [0.0001157, 2.15], [0.00011566, 1.32690645], [0.0001156, 0.8333045], [0.00011557, 100.0], [0.00011554, 0.9984421], [0.0001155, 2.91359307]], 'asks': [[0.00012725, 37.78664462], [0.00012728, 717.76343739], [0.00012748, 155.0], [0.00012758, 126.64805925], [0.00012769, 88.0], [0.00012776, 75.57328924], [0.0001278, 138.3097809], [0.00012786, 221.0], [0.00012798, 346.57033456], [0.00012815, 703.3901369], [0.00012847, 725.73340843], [0.00012879, 686.66571034], [0.00012887, 533.24742516], [0.000129, 120.19703592], [0.00012902, 20.0], [0.00012904, 1098.10028492], [0.00012908, 1098.14228492], [0.00012912, 166.36406548], [0.0001292, 0.00015479], [0.00012922, 8.279687], [0.00012943, 786.82035286], [0.00012959, 0.22], [0.00012975, 709.71417669], [0.00013, 1559.80523585], [0.00013006, 899.85404773], [0.00013019, 20.0], [0.00013037, 370.22074331], [0.0001305, 0.04796578], [0.0001307, 197.36471356], [0.00013089, 0.0021893], [0.000131, 705.82604028], [0.00013102, 303.72822758], [0.00013115, 8.51088711], [0.00013136, 20.0], [0.00013191, 870.97079443], [0.000132, 37.3679218], [0.0001323, 1.49676387], [0.00013254, 20.0], [0.00013259, 0.44], [0.00013298, 9.52336749], [0.000133, 1001.0], [0.00013309, 1.0], [0.00013311, 0.77], [0.00013318, 0.75], [0.00013333, 1.87187208], [0.00013348, 5.96827134], [0.0001335, 1.0], [0.00013373, 163.0], [0.00013374, 47.12002703], [0.000134, 3.24598162], [0.00013411, 6.17773803], [0.00013413, 134.72937237], [0.00013452, 28.9848961], [0.00013474, 6.17775026], [0.00013494, 20.0], [0.00013497, 5.0], [0.000135, 653.16842489], [0.00013504, 0.99], [0.00013538, 6.44405918], [0.00013593, 0.3], [0.00013601, 6.69614235], [0.00013605, 24.22952229], [0.00013615, 20.0], [0.00013659, 0.88], [0.0001366, 1.0], [0.00013665, 6.98496184], [0.00013667, 49.36781965], [0.0001367, 150.0], [0.00013689, 470.0], [0.000137, 5.53376673], [0.0001373, 6.98510865], [0.00013738, 20.0], [0.00013776, 487.0], [0.00013794, 6.98517847], [0.00013798, 5.0], [0.00013799, 15.0], [0.00013804, 0.99], [0.00013811, 0.99], [0.00013859, 8.17582418], [0.00013869, 500.0], [0.00013911, 1.22], [0.00013924, 8.17560684], [0.0001397, 750.0], [0.00013971, 0.31], [0.0001399, 8.86197898], [0.00013998, 250.0], [0.00013999, 6.5], [0.00014, 217.95714286], [0.00014056, 9.20862402], [0.00014104, 1.33], [0.00014122, 8.82774413], [0.00014145, 36.34278343], [0.00014165, 0.5267208], [0.00014188, 9.56922286], [0.00014249, 5.0], [0.00014255, 9.98192463], [0.0001427, 150.0], [0.00014322, 9.9822287], [0.0001439, 10.37248079], [0.000144, 3340.2325858]], 'timestamp': 1592149048891, 'datetime': '2020-06-14T15:37:28.891Z', 'nonce': None}
`

**Comment by kroitor** [comment]
@rmcxm we've added a few fixes for the URL and headers (those should work in 1.29.99, the new build will arrive in 15 minutes). 

The fix for `limit = 1` is also on the way.

**Comment by rmcxm** [comment]
thanks! you rock man!

**Comment by rmcxm** [comment]
fetch_balance has some issue

`>>> exchange.fetch_balance( )
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/ccxt/wavesexchange.py", line 1095, in fetch_balance
    hexSecret = base64.b16encode(self.base58_to_binary(self.secret))
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/ccxt/base/exchange.py", line 2097, in base58_to_binary
    result += Exchange.base58_decoder[s[i]]
KeyError: 'O'
>>> exchange.fetch_balance( params={'SYMBOL':'CDM24'})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/ccxt/wavesexchange.py", line 1095, in fetch_balance
    hexSecret = base64.b16encode(self.base58_to_binary(self.secret))
  File "/run/media/rishabh/NewVolume/fedora/envs/py37/lib/python3.7/site-packages/ccxt/base/exchange.py", line 2097, in base58_to_binary
    result += Exchange.base58_decoder[s[i]]
KeyError: 'O'
`

**Comment by kroitor** [comment]
@rmcxm thx for reporting, we're investigating now.
@frosty00 will followup on this.

**Comment by rmcxm** [comment]
also fetchClosedOrders etc... it is giving int to str error ...
`TypeError: can only concatenate str (not "int") to str
`
basically run a vanilla test for all apis.

**Comment by kroitor** [comment]
@rmcxm we will fix all of those shortly, thank you!

**Comment by frosty00** [comment]
@rmcxm

`secret` should be the base58 encoded privateKey from the website and the `apiKey` is the base58 encoded public key

**Comment by kroitor** [comment]
@rmcxm let us know if you have further difficulties with it after upgrading to 1.30.2
Thx to @frosty00 for the fixes!

**Comment by rmcxm** [comment]
Problem in private  api's :

`(py37) [rishabh@localhost ~]$ pip3 install --user  web3
Requirement already satisfied: web3 in /usr/local/lib/python3.7/site-packages (5.11.0)
Requirement already satisfied: websockets<9.0.0,>=8.1.0 in /usr/local/lib64/python3.7/site-packages (from web3) (8.1)
Requirement already satisfied: typing-extensions<4,>=3.7.4.1; python_version < "3.8" in /usr/local/lib/python3.7/site-packages (from web3) (3.7.4.2)
Requirement already satisfied: eth-utils<2.0.0,>=1.8.4 in /usr/local/lib/python3.7/site-packages (from web3) (1.9.0)
Requirement already satisfied: ipfshttpclient<1,>=0.4.12 in /usr/local/lib/python3.7/site-packages (from web3) (0.4.13.2)
Requirement already satisfied: eth-abi<3.0.0,>=2.0.0b6 in /usr/local/lib/python3.7/site-packages (from web3) (2.1.1)
Requirement already satisfied: eth-account<0.6.0,>=0.5.2 in /usr/local/lib/python3.7/site-packages (from web3) (0.5.2)
Requirement already satisfied: lru-dict<2.0.0,>=1.1.6 in /usr/local/lib64/python3.7/site-packages (from web3) (1.1.6)
Requirement already satisfied: hexbytes<1.0.0,>=0.1.0 in /usr/local/lib/python3.7/site-packages (from web3) (0.2.1)
Requirement already satisfied: eth-hash[pycryptodome]<1.0.0,>=0.2.0 in /usr/local/lib/python3.7/site-packages (from web3) (0.2.0)
Requirement already satisfied: protobuf<4,>=3.10.0 in /usr/local/lib64/python3.7/site-packages (from web3) (3.12.2)
Requirement already satisfied: jsonschema<4.0.0,>=3.0.0 in /usr/local/lib/python3.7/site-packages (from web3) (3.2.0)
Requirement already satisfied: eth-typing<3.0.0,>=2.0.0 in /usr/local/lib/python3.7/site-packages (from web3) (2.2.1)
Requirement already satisfied: requests<3.0.0,>=2.16.0 in /usr/lib/python3.7/site-packages (from web3) (2.22.0)
Requirement already satisfied: cytoolz<1.0.0,>=0.10.1; implementation_name == "cpython" in /usr/local/lib64/python3.7/site-packages (from eth-utils<2.0.0,>=1.8.4->web3) (0.10.1)
Requirement already satisfied: multiaddr>=0.0.7 in /usr/local/lib/python3.7/site-packages (from ipfshttpclient<1,>=0.4.12->web3) (0.0.9)
Requirement already satisfied: six in /usr/lib/python3.7/site-packages (from ipfshttpclient<1,>=0.4.12->web3) (1.12.0)
Requirement already satisfied: parsimonious<0.9.0,>=0.8.0 in /usr/local/lib/python3.7/site-packages (from eth-abi<3.0.0,>=2.0.0b6->web3) (0.8.1)
Requirement already satisfied: eth-keyfile<0.6.0,>=0.5.0 in /usr/local/lib/python3.7/site-packages (from eth-account<0.6.0,>=0.5.2->web3) (0.5.1)
Requirement already satisfied: bitarray<1.3.0,>=1.2.1 in /usr/local/lib64/python3.7/site-packages (from eth-account<0.6.0,>=0.5.2->web3) (1.2.2)
Requirement already satisfied: eth-keys!=0.3.2,<0.4.0,>=0.2.1 in /usr/local/lib/python3.7/site-packages (from eth-account<0.6.0,>=0.5.2->web3) (0.3.3)
Requirement already satisfied: eth-rlp<1,>=0.1.2 in /usr/local/lib/python3.7/site-packages (from eth-account<0.6.0,>=0.5.2->web3) (0.1.2)
Requirement already satisfied: rlp<2,>=1.0.0 in /usr/local/lib/python3.7/site-packages (from eth-account<0.6.0,>=0.5.2->web3) (1.2.0)
Requirement already satisfied: pycryptodome<4,>=3.6.6; extra == "pycryptodome" in /usr/local/lib64/python3.7/site-packages (from eth-hash[pycryptodome]<1.0.0,>=0.2.0->web3) (3.9.7)
Requirement already satisfied: setuptools in /usr/lib/python3.7/site-packages (from protobuf<4,>=3.10.0->web3) (40.8.0)
Requirement already satisfied: pyrsistent>=0.14.0 in /usr/local/lib64/python3.7/site-packages (from jsonschema<4.0.0,>=3.0.0->web3) (0.16.0)
Requirement already satisfied: importlib-metadata; python_version < "3.8" in /usr/local/lib/python3.7/site-packages (from jsonschema<4.0.0,>=3.0.0->web3) (1.6.1)
Requirement already satisfied: attrs>=17.4.0 in /usr/local/lib/python3.7/site-packages (from jsonschema<4.0.0,>=3.0.0->web3) (19.3.0)
Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /usr/lib/python3.7/site-packages (from requests<3.0.0,>=2.16.0->web3) (3.0.4)
Requirement already satisfied: idna<2.9,>=2.5 in /usr/lib/python3.7/site-packages (from requests<3.0.0,>=2.16.0->web3) (2.7)
Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/lib/python3.7/site-packages (from requests<3.0.0,>=2.16.0->web3) (1.25.7)
Requirement already satisfied: toolz>=0.8.0 in /usr/local/lib/python3.7/site-packages (from cytoolz<1.0.0,>=0.10.1; implementation_name == "cpython"->eth-utils<2.0.0,>=1.8.4->web3) (0.10.0)
Requirement already satisfied: netaddr in /usr/local/lib/python3.7/site-packages (from multiaddr>=0.0.7->ipfshttpclient<1,>=0.4.12->web3) (0.7.19)
Requirement already satisfied: base58 in /usr/local/lib/python3.7/site-packages (from multiaddr>=0.0.7->ipfshttpclient<1,>=0.4.12->web3) (2.0.1)
Requirement already satisfied: varint in /usr/local/lib/python3.7/site-packages (from multiaddr>=0.0.7->ipfshttpclient<1,>=0.4.12->web3) (1.0.2)
Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.7/site-packages (from importlib-metadata; python_version < "3.8"->jsonschema<4.0.0,>=3.0.0->web3) (3.1.0)

(py37) [rishabh@localhost ~]$ python
Python 3.7.4 (default, Aug 13 2019, 20:35:49) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ccxt
>>> exchange = ccxt.wavesexchange({'options': { }, 'urls': { 'api': {  'matcher': 'http://matcher.waves.exchange'}}, 'apiKey':'7FxcrEN94PKcC64VFht2uvfJtrGos3', 'secret':'G7qe7uqx57ATVUQu9KRvt4HfLheyATR5w'   })
>>> exchange.fetch_balance()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/rishabh/.local/lib/python3.7/site-packages/ccxt/wavesexchange.py", line 1049, in fetch_balance
    self.check_required_dependencies()
  File "/home/rishabh/.local/lib/python3.7/site-packages/ccxt/base/exchange.py", line 1815, in check_required_dependencies
    raise NotSupported("Web3 functionality requires Python3 and web3 package installed: https://github.com/ethereum/web3.py")
ccxt.base.errors.NotSupported: Web3 functionality requires Python3 and web3 package installed: https://github.com/ethereum/web3.py
>>> 
>>> 
`

**Comment by kroitor** [comment]
@rmcxm let's make sure that you have web3 properly installed first, try this:

```
(py37) [rishabh@localhost ~]$ python
```

```Python
import web3
```

**Comment by frosty00** [comment]
@rmcxm what is your ccxt version

**Comment by rmcxm** [comment]
`>>> ccxt.__version__
'1.30.2'
>>> import web3
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'web3'
`

**Comment by kroitor** [comment]
@rmcxm that error says, that if you did `pip3 install web3 --user` – it might have been installed into a different python site-packages directory than you expect, especially if you've had multiple versions of python installed on your system

Try

```
python3
```

```
import web3
```

↑ If that works, then it confirms the above.

In any case, whether it works or not, do cmd+f or ctrl+f or a linux find command for `python` or `python3.7` or `pip3` throughout your system – you will most likely find more than one python installed, and the web3 might have been installed into a wrong python directory.

In other words, this is your local python installation/configuration/path issue (i'm 99% sure). If you find more than one python – you might want to deinstall the wrong one and/or do a clean reinstall.

**Comment by kroitor** [comment]
@rmcxm alternatively, you can also try `pip3 install web3 ccxt` without the `--user` flag.

**Comment by rmcxm** [comment]
now stuck here : 

`>>> exchange.createLimitBuyOrder ('BOOC/CDM24', 2, 5000, { 'buyMatcherFee':'0.011' })
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/ccxt/base/exchange.py", line 580, in fetch
    response.raise_for_status()
  File "/run/media/rishabh/NewVolume/fedora/lib/python3.7/site-packages/requests/models.py", line 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: http://matcher.waves.exchange/matcher/orderbook

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/site-packages/ccxt/base/exchange.py", line 1792, in create_limit_buy_order
    return self.create_order(symbol, 'limit', 'buy', *args)
  File "/usr/local/lib/python3.7/site-packages/ccxt/wavesexchange.py", line 814, in create_order
    response = self.matcherPostMatcherOrderbook(body)
  File "/usr/local/lib/python3.7/site-packages/ccxt/base/exchange.py", line 462, in inner
    return entry(_self, **inner_kwargs)
  File "/usr/local/lib/python3.7/site-packages/ccxt/base/exchange.py", line 485, in request
    return self.fetch2(path, api, method, params, headers, body)
  File "/usr/local/lib/python3.7/site-packages/ccxt/base/exchange.py", line 481, in fetch2
    return self.fetch(request['url'], request['method'], request['headers'], request['body'])
  File "/usr/local/lib/python3.7/site-packages/ccxt/base/exchange.py", line 592, in fetch
    self.handle_errors(http_status_code, http_status_text, url, method, headers, http_response, json_response, request_headers, request_body)
  File "/usr/local/lib/python3.7/site-packages/ccxt/wavesexchange.py", line 1272, in handle_errors
    raise Exception(self.id + ' ' + message)
ccxt.base.errors.InvalidOrder: wavesexchange Required 0.011 WAVES as fee for this order, but given 0.003 WAVES
`

**Comment by kroitor** [comment]
@rmcxm this has been fixed in CCXT 1.30.24+, thanks for your feedback!

**Comment by rmcxm** [comment]
now this issue in fetch_balance:

`>>> ccxt.__version__
'1.30.25'
>>> exchange.fetch_balance()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/site-packages/ccxt/wavesexchange.py", line 1150, in fetch_balance
    result[code]['total'] = self.currency_from_precision(code, balance)
  File "/usr/local/lib/python3.7/site-packages/ccxt/wavesexchange.py", line 773, in currency_from_precision
    return self.from_wei(amount, self.currencies[currency]['precision'])
KeyError: '8KGUK1XSG11AC6X2MRBDFJNNJBZYWBQB51FVJRSWTVE5'
`

**Comment by kroitor** [comment]
@rmcxm will fix and resolve it shortly, thank you for your feedback!

**Comment by rmcxm** [comment]
some markets are not loading...

```
>>> exchange.load_markets()['WAVES/B00C']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'WAVES/B00C'

```

**Comment by frosty00** [comment]
not all the markets shown on the waves website are supported by their `fetchMarkets` API

**Comment by rmcxm** [comment]
then how to place order on 'BOOC/CDM24' pair?

**Comment by kroitor** [comment]
@rmcxm afaik, the Waves.Exchange matcher does not allow you do that via their API, correct me if I'm wrong.

**Comment by frosty00** [comment]
you can technically do it but you need to have the assetId for BOOC and CM24 which is not served via the current `fetchMarkets` endpoint
